pkg/gateway/backup/usecase/vald.go

```diff
@@ -0,0 +1,211 @@
+		return nil, errors.ErrInvalidBackupConfig
+	}
+
+	backupClientOptions := append(
+		cfg.Backup.Client.Opts(),
+		grpc.WithErrGroup(eg),
+	)
+
+	var obs observability.Observability
+	if cfg.Observability.Enabled {
+		obs, err = observability.NewWithConfig(cfg.Observability)
+		if err != nil {
+			return nil, err
+		}
+		backupClientOptions = append(
+			backupClientOptions,
+			grpc.WithDialOptions(
+				grpc.WithStatsHandler(metric.NewClientHandler()),
+			),
+		)
+	}
+
+	backup, err = service.NewBackup(
+		service.WithBackupAddr(cfg.Backup.Client.Addrs[0]),
+		service.WithBackupClient(
+			grpc.New(backupClientOptions...),
+		),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	if addrs := cfg.Client.Addrs; len(addrs) == 0 {
+		return nil, errors.ErrGRPCTargetAddrNotFound
+	}
+
+	v := handler.New(
+		handler.WithClient(client.New(
+			client.WithAddr(cfg.Client.Addrs[0]),
+			client.WithClient(grpc.New(cfg.Client.Opts()...)),
+		)),
+		handler.WithBackup(backup),
+		handler.WithErrGroup(eg),
+		handler.WithStreamConcurrency(cfg.Server.GetGRPCStreamConcurrency()),
+	)
+
+	grpcServerOptions := []server.Option{
+		server.WithGRPCRegistFunc(func(srv *grpc.Server) {
+			vald.RegisterValdServer(srv, v)
+		}),
+		server.WithPreStopFunction(func() error {
+			// TODO notify another gateway and scheduler
+			return nil
+		}),
+	}
+
+	if cfg.Observability.Enabled {
+		grpcServerOptions = append(
+			grpcServerOptions,
+			server.WithGRPCOption(
+				grpc.StatsHandler(metric.NewServerHandler()),
+			),
+		)
+	}
+
+	srv, err := starter.New(
+		starter.WithConfig(cfg.Server),
+		starter.WithREST(func(sc *config.Server) []server.Option {
+			return []server.Option{
+				server.WithHTTPHandler(
+					router.New(
+						router.WithHandler(
+							rest.New(
+								rest.WithVald(v),
+							),
+						),
+					),
+				),
+			}
+		}),
+		starter.WithGRPC(func(sc *config.Server) []server.Option {
+			return grpcServerOptions
+		}),
+		// TODO add GraphQL handler
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return &run{
+		eg:            eg,
+		cfg:           cfg,
+		server:        srv,
+		observability: obs,
+		backup:        backup,
+	}, nil
+}
+
+func (r *run) PreStart(ctx context.Context) error {
+	if r.observability != nil {
+		return r.observability.PreStart(ctx)
+	}
+	return nil
+}
+
+func (r *run) Start(ctx context.Context) (<-chan error, error) {
+	ech := make(chan error, 6)
+	var bech, sech, oech <-chan error
+	var err error
+	if r.observability != nil {
+		oech = r.observability.Start(ctx)
+	}
+	if r.backup != nil {
+		bech, err = r.backup.Start(ctx)
+		if err != nil {
+			close(ech)
+			return nil, err
+		}
+	}
+	sech = r.server.ListenAndServe(ctx)
+	r.eg.Go(safety.RecoverFunc(func() (err error) {
+		defer close(ech)
+		for {
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			case err = <-oech:
+			case err = <-bech:
+			case err = <-sech:
+			}
+			if err != nil {
+				select {
+				case <-ctx.Done():
+					return ctx.Err()
```

Comment:
do you need to returns the wrapped error or to output error log?

Reply:
I don't think so